File: Neural Network/NeuralNetwork.py
import numpy as np
from tqdm import tqdm

class NN:
    def __init__(self, X, Y, hidden_layers, randstate=None):
        if randstate:
            np.random.seed(randstate)
        self.X = X
        self.Y = Y
        self.hidden_layers = hidden_layers
        self.layers = [X.shape[1]] + hidden_layers + [Y.shape[1]]
        self.W = self.weights(self.layers)
        self.B = self.bias(self.layers)

    def weights(self, neurons):
        W = []
        for i in range(len(neurons) - 1):
            W.append(np.random.randn(neurons[i], neurons[i + 1]))
        return W

    def bias(self, neurons):
        B = []
        for i in range(1, len(neurons)):
            B.append(np.random.randn(neurons[i]))
        return B

    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))

    def relu(self, x):
        return (x + abs(x)) / 2

    def forward(self, X, W, B, activation=False):
        for i in range(len(W)):
            if activation:
                if i == len(W) - 1:
                    X = self.sigmoid(X @ W[i] + B[i])
                else:
                    X = self.relu(X @ W[i] + B[i])
            else:
                X = X @ W[i] + B[i]
        return X

    def modify_forward(self, X, W, B, h, hn, hi, hj=0, w_or_b="w", activation=False):
        if w_or_b == "w":
            W[hn][hi][hj] += h
        else:
            B[hn][hi] += h
        for i in range(len(W)):
            if activation:
                if i == len(W) - 1:
                    X = self.sigmoid(X @ W[i] + B[i])
                else:
                    X = self.relu(X @ W[i] + B[i])
            else:
                X = X @ W[i] + B[i]
        return X

    def MSE(self, Yt, Y):
        return ((Yt - Y) ** 2).mean()

    def grad(self, Y, X, W, B, h, lr, activation=False):
        # Weights
        for wt in range(len(W)):
            for wti in range(W[wt].shape[0]):
                for wtj in range(W[wt].shape[1]):
                    f1 = self.modify_forward(X, W, B, h, wt, wti, wtj, activation=activation)
                    f2 = self.modify_forward(X, W, B, -h, wt, wti, wtj, activation=activation)
                    gradient = (self.MSE(Y, f1) - self.MSE(Y, f2)) / (2 * h)
                    W[wt][wti][wtj] -= lr * gradient
        # Bias
        for b in range(len(B)):
            for bi in range(B[b].shape[0]):
                f1 = self.modify_forward(X, W, B, h, b, bi, w_or_b="b", activation=activation)
                f2 = self.modify_forward(X, W, B, -h, b, bi, w_or_b="b", activation=activation)
                gradient = (self.MSE(Y, f1) - self.MSE(Y, f2)) / (2 * h)
                B[b][bi] -= lr * gradient
        return W, B

    def backpropagation(self, h, lr, epochs, activation=False):
        self.activation = True
        loss_history = []
        loss_history.append(self.MSE(self.Y, self.forward(self.X, self.W, self.B, activation)))
        for _ in tqdm(range(epochs)):
            self.W, self.B = self.grad(self.Y, self.X, self.W, self.B, h, lr, activation)
            loss_history.append(self.MSE(self.Y, self.forward(self.X, self.W, self.B, activation)))
        print(
            f"\nInitial Loss: {loss_history[0]}\nFinal Loss: {loss_history[-1]}\nMin Loss: {min(loss_history)}\n"
        )
        return loss_history

    def predict(self, X):
        for i in range(len(self.W)):
            if self.activation:
                if i == len(self.W) - 1:
                    X = self.sigmoid(X @ self.W[i] + self.B[i])
                else:
                    X = self.relu(X @ self.W[i] + self.B[i])
            else:
                X = X @ self.W[i] + self.B[i]
        return X

File: Neural Network in Rust/src/neural_network.rs
use std::io::Write;
use crate::matrix::{rand_matrix, dot, rand_vec, add_arr2_arr1, mse};

#[allow(unused)]
pub struct NN{
    x:Vec<Vec<f32>>,
    y:Vec<Vec<f32>>,
    hidden_layers:Vec<usize>,
    layers:Vec<usize>,
    w:Vec<Vec<Vec<f32>>>,
    b:Vec<Vec<f32>>,
    activation:bool
}

fn weights(neurons:&Vec<usize>) -> Vec<Vec<Vec<f32>>> {
    let mut w = vec![];
    for i in 0..(neurons.len() - 1) {
        w.push(rand_matrix(neurons[i], neurons[i + 1], 0.0, 1.0));
    }
    return w;
}

fn bias(neurons:&Vec<usize>) -> Vec<Vec<f32>> {
    let mut b = Vec::new();
    for i in 1..neurons.len() {
        b.push(rand_vec(neurons[i]));
    }
    return b
}

impl NN {
    pub fn new(x:Vec<Vec<f32>>, y:Vec<Vec<f32>>, hl:Vec<usize>) -> Self {
        let mut layers = hl.clone();
        layers.insert(0, x[0].len());
        layers.push(y[0].len());
        Self {x,y,hidden_layers:hl,layers:layers.clone(),
            w:weights(&layers),b:bias(&layers),
            activation:false}
    }

    #[allow(unused)]
    pub fn forward(&self) -> Vec<Vec<f32>> {
        let mut z = self.x.clone();
        for i in 0..self.w.len() {
            z = add_arr2_arr1(&dot(&z, &self.w[i]), &self.b[i]);
        }
        return z;
    }

    #[allow(unused)]
    fn modify_forward(&self, w:&mut Vec<Vec<Vec<f32>>>, b:&mut Vec<Vec<f32>>, h:f32, hn:usize, hi:usize, hj:usize, w_or_b:char) -> Vec<Vec<f32>>{
        if w_or_b == 'w'{
            w[hn][hi][hj] += h
        }
        else {
            b[hn][hi] += h
        }

        let mut z = self.x.clone();
        for i in 0..w.len() {
            z = add_arr2_arr1(&dot(&z, &w[i]), &b[i]);
        }
        return z;
    }

    #[allow(unused)]
    fn grad(&self, h:f32, lr:f32) -> (Vec<Vec<Vec<f32>>>, Vec<Vec<f32>>) {
        // Weights
        let mut w = self.w.clone();
        let mut b = self.b.clone();
        for wt in 0..w.len() {
            for wti in 0..w[wt].len() {
                for wtj in 0..w[wt][0].len() {
                    let f1 = self.modify_forward(&mut w, &mut b, h, wt, wti, wtj, 'w');
                    let f2 = self.modify_forward(&mut w, &mut b, -h, wt, wti, wtj, 'w');
                    let gradient = (mse(&self.y, &f1) - mse(&self.y, &f2)) / (2.0 * h);
                    w[wt][wti][wtj] -= lr * gradient;
                }
            }
        }

        // Bias
        for i in 0..b.len() {
            for j in 0..b[i].len() {
                let f1 = self.modify_forward(&mut w, &mut b, h, i, j, 0, 'b');
                let f2 = self.modify_forward(&mut w, &mut b, -h, i, j, 0, 'b');
                let gradient = (mse(&self.y, &f1) - mse(&self.y, &f2)) / (2.0 * h);
                b[i][j] -= lr * gradient;
            }
        }
        return (w, b);
    }

    #[allow(unused)]
    pub fn backpropagation(&mut self, h:f32, lr:f32, epochs:usize) -> Vec<f32> {
        let mut loss_history = vec![];
        loss_history.push(mse(&self.y, &self.forward()));
        for e in 0..epochs {
            (self.w, self.b) = self.grad(h, lr);
            let l = mse(&self.y, &self.forward());
            loss_history.push(l);
            print!("\rEpoch: {:4} ({}%), Loss: {:.5} ({:.2}%)  ", e+1, (e+1)*100/epochs, l, (l-loss_history[0])*100.0/loss_history[0]);
            std::io::stdout().flush().unwrap();
        }
        println!("\nInitial Loss: {}\nFinal Loss: {}\n", 
        loss_history[0], 
        loss_history[loss_history.len()-1]);
        return loss_history;
    }

    #[allow(unused)]
    pub fn predict(&self, x:&Vec<Vec<f32>>) -> Vec<Vec<f32>> {
        let mut z = x.clone();
        for i in 0..self.w.len() {
            z = add_arr2_arr1(&dot(&z, &self.w[i]), &self.b[i]);
        }
        return z;
    }
}